### PR TITLE
optimize sort by inlining comparison functions

### PR DESCRIPTION
**Pull request by xenu**
This makes special-cased forms such as `sort { $b <=> $a }`
even faster.

Also, since this commit removes `PL_sort_RealCmp`, it fixes the
issue with nested sort calls mentioned in #16129

PS. Because of b578cabd8b430a108c01d78e451753f3d2c10a47 commit (normalize indentation), I recommend using "hide whitespace changes" when reviewing.

### CODE REVIEW

**Comment by xenu** [comment]
Some benchmarks ([the source code is here](https://github.com/Perl/perl5/files/4278590/bench_code.txt)), "fastblead" is the one with my changes:
```
Key:
    Ir   Instruction read
    Dr   Data read
    Dw   Data write
    COND conditional branches
    IND  indirect branches
    _m   branch predict miss
    _m1  level 1 cache miss
    _mm  last cache (e.g. L3) miss
    -    indeterminate percentage (e.g. 1/0)

The numbers represent relative counts per loop iteration, compared to
/home/xenu/blead/bin/perl5.31.10 at 100.0%.
Higher is better: for example, using half as many instructions gives 200%,
while using twice as many gives 50%.

nsort_asc
numeric sort ascending

       /home/xenu/blead/bin/perl5.31.10 /home/xenu/fastblead/bin/perl5.31.10
       -------------------------------- ------------------------------------
    Ir                           100.00                               115.45
    Dr                           100.00                               113.63
    Dw                           100.00                               139.99
  COND                           100.00                               100.00
   IND                           100.00                           1000090.00

COND_m                           100.00                                88.89
 IND_m                           100.00                               100.00

 Ir_m1                           100.00                               100.00
 Dr_m1                           100.00                               100.00
 Dw_m1                           100.00                               100.00

 Ir_mm                           100.00                               100.00
 Dr_mm                           100.00                               100.00
 Dw_mm                           100.00                               100.00

nsort_desc
numeric sort descending

       /home/xenu/blead/bin/perl5.31.10 /home/xenu/fastblead/bin/perl5.31.10
       -------------------------------- ------------------------------------
    Ir                           100.00                               119.64
    Dr                           100.00                               120.00
    Dw                           100.00                               153.32
  COND                           100.00                               104.97
   IND                           100.00                           2000080.00

COND_m                           100.00                               100.00
 IND_m                           100.00                               100.00

 Ir_m1                           100.00                               100.00
 Dr_m1                           100.00                               100.00
 Dw_m1                           100.00                               100.00

 Ir_mm                           100.00                               100.00
 Dr_mm                           100.00                               100.00
 Dw_mm                           100.00                               100.00

ssort_asc
string sort ascending

       /home/xenu/blead/bin/perl5.31.10 /home/xenu/fastblead/bin/perl5.31.10
       -------------------------------- ------------------------------------
    Ir                           100.00                               100.00
    Dr                           100.00                               101.25
    Dw                           100.00                               100.00
  COND                           100.00                               100.00
   IND                           100.00                               150.00

COND_m                           100.00                               100.00
 IND_m                           100.00                               100.00

 Ir_m1                           100.00                               100.00
 Dr_m1                           100.00                               100.00
 Dw_m1                           100.00                               100.00

 Ir_mm                           100.00                               100.00
 Dr_mm                           100.00                               100.00
 Dw_mm                           100.00                               100.00

ssort_desc
string sort descending

       /home/xenu/blead/bin/perl5.31.10 /home/xenu/fastblead/bin/perl5.31.10
       -------------------------------- ------------------------------------
    Ir                           100.00                               103.35
    Dr                           100.00                               106.23
    Dw                           100.00                               106.06
  COND                           100.00                               100.00
   IND                           100.00                               200.00

COND_m                           100.00                                99.99
 IND_m                           100.00                               100.00

 Ir_m1                           100.00                               100.00
 Dr_m1                           100.00                               100.00
 Dw_m1                           100.00                               100.00

 Ir_mm                           100.00                               100.00
 Dr_mm                           100.00                               100.00
 Dw_mm                           100.00                               100.00

AVERAGE

       /home/xenu/blead/bin/perl5.31.10 /home/xenu/fastblead/bin/perl5.31.10
       -------------------------------- ------------------------------------
    Ir                           100.00                               109.00
    Dr                           100.00                               109.82
    Dw                           100.00                               120.87
  COND                           100.00                               101.20
   IND                           100.00                               342.82

COND_m                           100.00                                96.97
 IND_m                           100.00                               100.00

 Ir_m1                           100.00                               100.00
 Dr_m1                           100.00                               100.00
 Dw_m1                           100.00                               100.00

 Ir_mm                           100.00                               100.00
 Dr_mm                           100.00                               100.00
 Dw_mm                           100.00                               100.00
```

**Comment by xenu** [comment]
Here are some non-scientific benchmarks of numeric sort ([source code](https://github.com/Perl/perl5/files/4278963/num_sort.txt)):

```
> time ~/blead/bin/perl5.31.10 num_sort.pl

real    0m2.988s
user    0m2.875s
sys     0m0.047s
> time ~/blead/bin/perl5.31.10 num_sort.pl

real    0m2.984s
user    0m2.922s
sys     0m0.047s
> time ~/blead/bin/perl5.31.10 num_sort.pl

real    0m2.982s
user    0m2.953s
sys     0m0.000s
```

```
> time ~/fastblead/bin/perl5.31.10 num_sort.pl

real    0m2.485s
user    0m2.453s
sys     0m0.016s
> time ~/fastblead/bin/perl5.31.10 num_sort.pl

real    0m2.481s
user    0m2.406s
sys     0m0.047s
> time ~/fastblead/bin/perl5.31.10 num_sort.pl

real    0m2.519s
user    0m2.438s
sys     0m0.047s
```

and reverse numeric sort ([source code](https://github.com/Perl/perl5/files/4278962/reverse_num_sort.txt)):

```
> time ~/blead/bin/perl5.31.10 reverse_num_sort.pl

real    0m3.198s
user    0m3.141s
sys     0m0.016s
> time ~/blead/bin/perl5.31.10 reverse_num_sort.pl

real    0m3.206s
user    0m3.172s
sys     0m0.016s
> time ~/blead/bin/perl5.31.10 reverse_num_sort.pl

real    0m3.197s
user    0m3.156s
sys     0m0.016s
```

```
> time ~/fastblead/bin/perl5.31.10 reverse_num_sort.pl

real    0m2.578s
user    0m2.516s
sys     0m0.047s
> time ~/fastblead/bin/perl5.31.10 reverse_num_sort.pl

real    0m2.577s
user    0m2.531s
sys     0m0.016s
> time ~/fastblead/bin/perl5.31.10 reverse_num_sort.pl

real    0m2.566s
user    0m2.484s
sys     0m0.063s
```

As you can see, on my machine the optimized numeric sort is ~15% faster and reverse numeric sort is ~20% faster.

**Comment by xenu** [comment]
I force pushed to improve a comment in `embed.fnc`.

File: proto.h [hunk 1]
```diff
@@ -684,9 +684,11 @@ PERL_CALLCONV_NO_RET void	Perl_croak_caller(const char* pat, ...)
 			__attribute__format__null_ok__(__printf__,1,2);
 #define PERL_ARGS_ASSERT_CROAK_CALLER
 
+#ifndef PERL_NO_INLINE_FUNCTIONS
```
Comment (khwilliamson) [inline proto.h:687]: I presume this change is due to a bug fix in the regen file

File: regen/embed.pl [hunk 2]
```diff
@@ -132,14 +140,14 @@ sub open_print_header {
 
 	die_at_end "For '$plain_func', M flag requires p flag"
 					    if $flags =~ /M/ && $flags !~ /p/;
-	die_at_end "For '$plain_func', C flag requires one of [pim] flag"
-					    if $flags =~ /C/ && $flags !~ /[ibmp]/;
-	die_at_end "For '$plain_func', X flag requires p or i flag"
-					    if $flags =~ /X/ && $flags !~ /[ip]/;
+	die_at_end "For '$plain_func', C flag requires one of [bmIip] flags"
```
Comment (khwilliamson) [inline regen/embed.pl]: This is an unrelated bug fix, that the diagnostic didn't match the check.  I don't know that it matters here, but normally this change should go in a separate commit, so that if this one, for whatever reason, gets reverted, the unrelated bug fix stays.

Comment (khwilliamson) [reply regen/embed.pl]: Also, should there be a check that i and I aren't both specified?  Maybe other illegal combos?

**Review by khwilliamson — APPROVED** [review]
LGTM

**Review by khwilliamson — APPROVED** [review]

**Review by khwilliamson — APPROVED** [review]

**Review by khwilliamson — APPROVED** [review]

File: proto.h (hunk 1 above)
Comment (xenu) [reply proto.h:687]: yep, it's a (unintended) side-effect of replacing `eq 'PERL_STATIC_INLINE'` with `=~ /PERL_STATIC(?:_FORCE)?_INLINE/`

File: regen/embed.pl (hunk 2 above)
Comment (xenu) [reply regen/embed.pl]: I think I agree, I'll move it to a separate commit. And yes, adding that check makes sense too.

File: pp_sort.c [hunk 3]
```diff
@@ -179,7 +175,7 @@ typedef SV * gptr;              /* pointers in our lists */
 */
 
 
-static IV
+PERL_STATIC_FORCE_INLINE IV __attribute__always_inline__
```
Comment (khwilliamson) [inline pp_sort.c:178]: I think here and another place, the __attribute__always_inline__ is a relic of an earlier version, and is generated by the macro.  Most functions declared as such in this file don't have an __attribute__ specified explicitly

**Comment by dur-randir** [comment]
I observe ~5% overall improvement, which is very nice for such a used function.

File: pp_sort.c (hunk 3 above)
Comment (xenu) [reply pp_sort.c:178]: I've added `sortsv_flags_impl` to `embed.fnc` and removed the attribute from its definition but I would rather not do the same for `dynprep()`.

`dynprep()` takes arguments of types that exist only inside `pp_sort.c`, so I'd have to either move the typedefs to `perl.h` (which is an obfuscation) or remove them (which is beyond the scope of this PR).

**Comment by hvds** [comment]
@xenu  I note that this is now generating `warning: unused parameter 'flags'` on the renamed `S_sortsv_flags_impl`. It is not immediately clear to me whether it just wants a PERL_UNUSED_ARG or whether the interface should actually remove flags from `_impl` and all the specific sort functions.

**Comment by jkeenan** [comment]
> @xenu I note that this is now generating `warning: unused parameter 'flags'` on the renamed `S_sortsv_flags_impl`. It is not immediately clear to me whether it just wants a PERL_UNUSED_ARG or whether the interface should actually remove flags from `_impl` and all the specific sort functions.

As reported in https://github.com/Perl/perl5/issues/17632
